Packs/DarktraceEventCollector/Integrations/DarktraceEventCollector/DarktraceEventCollector.py

```diff
@@ -0,0 +1,283 @@
+    public_token, private_token = tokens
+    query_string = f'?{json.dumps(query_data)}' if is_json else f'?{stringify_data(query_data)}' if query_data else ''
+
+    return hmac.new(
+        private_token.encode('ASCII'),
+        f'{query_uri}{query_string}\n{public_token}\n{date}'.encode('ASCII'),
+        hashlib.sha1,
+    ).hexdigest()
+
+
+def filter_events(events: List[Dict[str, Any]], last_fetched_pid: int, max_fetch: int) -> List[Dict[str, Any]]:
+    """Filters events by ascending pbid and max_fetch"""
+    for index, event in enumerate(events):
+        if event.get('pbid', 0) > last_fetched_pid:
+            return events[index:index + max_fetch]
+    return []
+
+
+def add_time_field(events: List[Dict[str, Any]]) -> List[Dict[str, Any]]:
+    """Adds time field to the events"""
+    for event in events:
+        event['_time'] = timestamp_to_datestring(event['creationTime'])
+    return events
+
+
+def convert_to_timestamp(date: datetime | None) -> int:
+    """Converts datetime to timestamp"""
+    if date:
+        if isinstance(date, datetime):
+            return int(date.timestamp())
+        elif isinstance(date, int):
+            return int(date)
+    return 0
+
+
+"""*****COMMAND FUNCTIONS****"""
+
+
+def test_module(client: Client, first_fetch_time: int) -> str:
+    """
+     Returning 'ok' indicates that the integration works like it is supposed to. Connection to the service is successful.
+    """
+    try:
+        client.get_events(start_time=first_fetch_time, end_time=int(datetime.now().timestamp()))
+    except DemistoException as e:
+        if 'Forbidden' in str(e):
+            return 'Authorization Error: make sure API Key is correctly set'
+        else:
+            raise e
+    return 'ok'
+
+
+def fetch_events(client: Client, max_fetch: int, last_run: Dict[str, Any],
+                 start_time: int, end_time: int) -> \
+        tuple[List[Dict[str, Any]], Dict[str, Any]]:
+    """
+       Fetches events from Darktrace API.
+    """
+    start_time = last_run.get('last_fetch_time', start_time)
+    demisto.debug(f'Getting events from: {timestamp_to_datestring(start_time)}, till: {timestamp_to_datestring(end_time)}')
+    retrieve_events = client.get_events(start_time, end_time)
+    demisto.debug(f'Fetched {len(retrieve_events)} events.')
+
+    # filtering events
+    retrieve_events = filter_events(retrieve_events, int(last_run.get('last_fetch_pid', 0)), max_fetch)
+    demisto.debug(f'Limiting to {len(retrieve_events)} events.')
+
+    # setting last run object
+    if retrieve_events:
+        # extracting last fetch time and last fetched events.
+        last_fetch_time = retrieve_events[-1].get('time')
+        last_fetched_pbid = retrieve_events[-1].get('pbid')
+        demisto.debug(f'Setting last run to pbid: {last_fetched_pbid} time:{timestamp_to_datestring(last_fetch_time)}')
+        last_run = {'last_fetch_time': retrieve_events[-1].get('creationTime'),
```

Comment:
let's say X is last event I pulled, with creation time Y. I store Y in the last run, then use it in the query of the next run. Won't I pull X (and other events that occurred at time Y) again? If so - make sure to dedup

Reply:
We are taking care of it in the `filter_events` command. (see [this](https://github.com/demisto/content/pull/27542/files#diff-12d5c6c2bf9cc0f8ec33c03049f0e5ca57f74b9d4bc7a0855c42efc098b741ab:~:text=if%20event.get(%27pbid%27%2C%200)%20%3E%20last_fetched_pid%3A) line)